### ISSUE TITLE
Towards 0.2

### ISSUE THREAD

**Issue by yianni** [strands-project/strands_qsr_lib#137]
Help needed.

[This branch](https://github.com/yianni/strands_qsr_lib/tree/towards-0.2) includes some changes that we have been discussing, mainly:
* restructuring the code to make it simpler for devs
* making --future the only option

Still pending are:
* @cdondrup QTCBS family has not been restructured; I have provided the abstractmethods that need to be implemented; can you please restructure it? I will provide more details below.
* @pet1330 Can you change the RCC relations according to PR #128 Sorry but it was not merged when I pulled master; shouldn't take long.

Some notes:
* `make` function is replaced by `make_world_qsr_trace`. This is the main function that a dev should implement and has a lot of the previous overhead removed.
* The overhead now is integrated in functions:
 * `_process_qsrs_for`
 * `_process_qsr_parameters_from_request_parameters`
 * `_postprocess_world_qsr_trace`
 * `_init_qsrs_for_default`
There are also new abstract classes:
* `QSR_Monadic_Abstractclass`
* `QSR_Dyadic_Abstractclass`
that also handle some of the common overhead.

PR that needs rejecting:
* #136 as default behaviour of dyadic relations is to return all possible combinations (i.e. including mirrors) see also file `combinations_and_permutations`
* #131 
* #128

Also I suggest blocking the remaining PRs till we get this sorted first.

Footnotes:
* argd and argprod are bugged, working on it.
* code far from cleaned up but can be done in parallel.
* `qsrs_for` not fully tested but with the new way it is much easier to change and debug to what we finally decide.

Any questions shoot.

Many thanks.

**Comment by yianni** [comment]
Confirm you agree with PR rejections.

**Comment by yianni** [comment]
Ah yes and for examples see `MOS` or `RCC2`.

**Comment by cdondrup** [comment]
I have  few comments, sadly I cannot make inline comments, so please bare with me:

* [qsrlib](https://github.com/yianni/strands_qsr_lib/blob/8ff3c5806075d7b4133ccbf74d02dcf02b272552/qsr_lib/src/qsrlib/qsrlib.py#L156-L157) the previous try and except should be faster than an if. The if is executed every time even though the default case will be that the qsr does exist. The except block is only executed when the exception occurs.

* qsr_abstractclass 
 * [postprocess](https://github.com/yianni/strands_qsr_lib/blob/8ff3c5806075d7b4133ccbf74d02dcf02b272552/qsr_lib/src/qsrlib_qsrs/qsr_abstractclass.py#L89-L91): why is this abstract. The normal implementation will always be:

  ```python
def _postprocess_world_qsr_trace(self, world_qsr_trace, world_trace, world_trace_timestamps, req_params, qsr_params):
        return world_qsr_trace
  ```

  why burden the dev with copying this? I only know of qtc that would currently need this being overridden.
 * [handle_future](https://github.com/yianni/strands_qsr_lib/blob/8ff3c5806075d7b4133ccbf74d02dcf02b272552/qsr_lib/src/qsrlib_qsrs/qsr_abstractclass.py#L146-L148): Just remove this. We want to get rid of deprecation, not include any more.
  * [parameters](https://github.com/yianni/strands_qsr_lib/blob/8ff3c5806075d7b4133ccbf74d02dcf02b272552/qsr_lib/src/qsrlib_qsrs/qsr_abstractclass.py#L85-L87): This could have a standard implementation:

  ```python
def _process_qsr_parameters_from_request_parameters(self, req_params, **kwargs):
        try:
            qsr_params = req_params["dynamic_args"][self._unique_id]
        except KeyError:
            try:
                qsr_params = req_params["dynamic_args"]["for_all_qsrs"]
            except TypeError:
                pass
        except TypeError:
            pass
        return qsr_params
  ```
  and just return the dictionary?!


Regarding rejection of QSRs:

* #136 This returns all combinations as well but makes sure that they are only computed once. Do you do that as well? Couldn't find it on first glance.
* #131 Is yours anyway, no opinion.
* #128 should imho be merged now and then @yianni rebases his branch and makes the necessary adjustments

Regarding the blocking of PRs, that depends on who you want to do the work. The creator of the PR by giving your work precedence or yourself by merging the PRs now and then you making the necessary adjustments to resolve conflicts. Apart from sparing me some work, I also think that you know your changes the best and it would therefore make sense that you make them instead of us trying to find out what changed and then rebasing all our PRs.

**Comment by cdondrup** [comment]
BTW: Have you seen this: https://github.com/strands-project/lamor15/issues/1#issuecomment-133329040 ?

**Comment by yianni** [comment]
See below:

> I have few comments, sadly I cannot make inline comments, so please bare with me:
> 
> qsrlib the previous try and except should be faster than an if. The if is executed every time even though the default case will be that the qsr does exist. The except block is only executed when the exception occurs.
This if hardly puts any overload, but that's fine, will take it off. The try/except makes debugging a pain... the error can be anything and different than the message. Might as well leave it unprotected.
> 
> qsr_abstractclass
> 
> postprocess: why is this abstract.
Agree. Non-abstract is fine.
> 
> handle_future: Just remove this. We want to get rid of deprecation, not include any more.
Code is being cleaned up, this is not used anywhere at the moment.

> parameters: This could have a standard implementation:
> def _process_qsr_parameters_from_request_parameters(self, req_params, **kwargs):
>     try:
>         qsr_params = req_params["dynamic_args"][self._unique_id]
>     except KeyError:
>         try:
>             qsr_params = req_params["dynamic_args"]["for_all_qsrs"]
>         except TypeError:
>             pass
>     except TypeError:
>         pass
>     return qsr_params
> and just return the dictionary?!
Don't like this for many reasons, it does not allow pick-and-mix between the two namespaces, hence might as well force only the QSR namespace as the for_all_qsrs has no use then, creating some overhead for common attributes.

> Regarding rejection of QSRs:
> 
> #136 This returns all combinations as well but makes sure that they are only computed once. Do you do that as well? Couldn't find it on first glance.
What do you mean? If it does so, then I am very much against it because this way forces that all data should be present. Such change needs to be clear and discussed beforehand.

> #131 Is yours anyway, no opinion.
Still doesn't matter who makes the PR.

> #128 should imho be merged now and then @yianni rebases his branch and makes the necessary adjustments
Can do that also.

> Regarding the blocking of PRs, that depends on who you want to do the work. The creator of the PR by giving your work precedence or yourself by merging the PRs now and then you making the necessary adjustments to resolve conflicts. Apart from sparing me some work, I also think that you know your changes the best and it would therefore make sense that you make them instead of us trying to find out what changed and then rebasing all our PRs.

Don't mind either way.

**Comment by cdondrup** [comment]
>> I have few comments, sadly I cannot make inline comments, so please bare with me:

>> qsrlib the previous try and except should be faster than an if. The if is executed every time even though the default case will be that the qsr does exist. The except block is only executed when the exception occurs.

> This if hardly puts any overload, but that's fine, will take it off. The try/except makes debugging a pain... the error can be anything and different than the message. Might as well leave it unprotected.

It only looks for a `KeyError` but I agree that the key error might also occur down the line. Fell free to change it if you like.

>> parameters: This could have a standard implementation:
>> def _process_qsr_parameters_from_request_parameters(self, req_params, **kwargs):
>> try:
>> qsr_params = req_params["dynamic_args"][self._unique_id]
>> except KeyError:
>> try:
>> qsr_params = req_params["dynamic_args"]["for_all_qsrs"]
>> except TypeError:
>> pass
>> except TypeError:
>> pass
>> return qsr_params
>> and just return the dictionary?!

> Don't like this for many reasons, it does not allow pick-and-mix between the two namespaces, hence might as well force only the QSR namespace as the for_all_qsrs has no use then, creating some overhead for common attributes.

This should do the trick and take the global ones if there are no namespace ones

``` python
def _process_qsr_parameters_from_request_parameters(self, req_params, **kwargs):
    qsr_params = {}
    try:
        for k,v in req_params["dynamic_args"]["for_all_qsrs"]:
            qsr_params[k] = v
    except KeyError:
        pass
    try:
        for k,v in req_params["dynamic_args"][self._unique_id]:
            qsr_params[k] = v
    except KeyError:
        pass
    return qsr_params
```

>> #136 This returns all combinations as well but makes sure that they are only computed once. Do you do that as well? Couldn't find it on first glance.

> What do you mean? If it does so, then I am very much against it because this way forces that all data should be present. Such change needs to be clear and discussed beforehand.

The argd previously created only QSRs for one object combination and excluded the inverse one. This resulted in `o1,o2` having a `argd` value and `o2,o1` not having one which is very non deterministic in nature. The PR just removes the check where inverse object combinations are discared and then calculates `argd` for `o1,o2` and looks it up for `o2,o1`. I don't see a problem there as this is only used if no `qsrs_for` is given.

>> #131 Is yours anyway, no opinion.

> Still doesn't matter who makes the PR. Feel free to comment.

I'm just not sure if the new branch resolves this issue a different a way so I relied on your judgment ;) Still I think that the inconsistent way of implementation used in #131 should never be merged but but I already said so in #114

**Comment by yianni** [comment]
Sceptical about the one below as allows the user to give its own names and break things. Anyway can also be changed later; as I am trying to get the core of the 0.2 out ASAP. Such improvements can be done later.
> This should do the trick and take the global ones if there are no namespace ones
```python
def _process_qsr_parameters_from_request_parameters(self, req_params, **kwargs):
    qsr_params = {}
    try:
        for k,v in req_params["dynamic_args"]["for_all_qsrs"]:
            qsr_params[k] = v
    except KeyError:
        pass
    try:
        for k,v in req_params["dynamic_args"][self._unique_id]:
            qsr_params[k] = v
    except KeyError:
        pass
    return qsr_params
```

**Comment by yianni** [comment]
@cdondrup OK, the latest commit on my towards-0.2 branch is missing the QTCS family to be brought up to speed. Can you have a look please? I will carry on with some other things in the meantime but I won't touch the QTCS* files to avoid conflicts. If you can do the same and stay out of the rest of the files other than the QTCS* it would be great :P.

**Comment by yianni** [comment]
Ah yes and ask me if you need something.... but in the end of the day all needed should be the `make_world_qsr_trace` and `process_qsr_parameters...` functions. Look MOS for example.

**Comment by cdondrup** [comment]
Will do but might not be able to before tomorrow.

**Comment by yianni** [comment]
That's fine. Thanks.

**Comment by cdondrup** [comment]
I'll leave a few suggestions here while which I came across while altering QTC:

* Set `req_params = req_params["dynamic_args"]`
* Set `_for_all_qsrs = "for_all_qsrs"` in abstract class so one can use `req_params[self._for_all_qsrs]` to get the global ones.
* Check if the objects are available in both timestamps before running calculations:

 ```python
if set(world_state_now.objects.keys()) != set(world_state_previous.objects.keys()):
                ret.add_empty_world_qsr_state(t)
                continue # Objects have to be present in both timestamps
```

* custom_checks_qsrs_for still necessary? You state this in the code:

 ```python
    def custom_checks_for_qsrs_for(self, qsrs_for, error_found):
        """Custom checks of the qsrs_for field

        :param qsrs_for: list of strings and/or tuples for which QSRs will be computed
        :param error_found: if an error was found in the qsrs_for that violates the QSR rules
        :return: qsrs_for, error_found
        """
        raise DeprecationWarning("No longer needed")
 ```

But in `qsr_abstractclass` line 114 we find this: `qsrs_for_ret = self.custom_checks_for_qsrs_for(qsrs_for_ret)`

OK, I'll open a PR to your branch @yianni 
QTC now works for me.

**Comment by yianni** [comment]
I'll merge the PR and have a look later... Some things have changed since then.

**Comment by cdondrup** [comment]
Just saw it... have to rebase first, got conflicts.
